src/ocean/meta/traits/Basic.d

```diff
@@ -0,0 +1,424 @@
+/*******************************************************************************
+
+    Basic traits allowing to distinguish various types between each other. Any
+    more convoluted traits dedicated to specific type kinds should go in
+    dedicated modules.
+
+    Copyright:
+        Copyright (C) 2017 Sociomantic Labs GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.meta.traits.Basic;
+
+import ocean.meta.types.Qualifiers : Unqual, Const;
+
+/*******************************************************************************
+
+    Types are the types from which one or multiple other types cannot be
+    derived from using the ``is()`` expression or corresponding template
+    type parameter specialisation. The following types are not primitive:
+     - arrays (static, dynamic and associative) and pointers,
+     - classes structs and unions,
+     - delegates, function pointers and functions (function pointer base types),
+     - enums and typedefs.
+
+    All other, including arithmetic and character types are primitive. Each
+    primitive type is represented by a D keyword.
+    ``void`` is a primitive type. Imaginary and complex numbers are considered
+    primitive types, too, which may be subject to discussion.
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is a primitive type
+
+*******************************************************************************/
+
+///
+public template isPrimitiveType ( T )
+{
+    const isPrimitiveType =
+           is(Unqual!(T) == void)
+        || is(Unqual!(T) == bool)
+        || isIntegerType!(T)
+        || isCharType!(T)
+        || isFloatingPointType!(T);
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is one of supported string element types
+
+*******************************************************************************/
+
+public template isCharType ( T )
+{
+    const bool isCharType =
+           is(Unqual!(T) == char)
+        || is(Unqual!(T) == wchar)
+        || is(Unqual!(T) == dchar);
+}
+
+///
+unittest
+{
+    static assert ( isCharType!(wchar));
+    static assert ( isCharType!(Const!(char)));
+    static assert (!isCharType!(byte));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is a signed integer type
```

Comment:
The docs for all of these traits should say something like `built-in signed integer type`, so that it's clear that they don't include synthetic types like `BigInt`, `CustomFloat`.